### PR TITLE
[Security Hardened Shoot Cluster] Rule 1000 Implementation

### PR DESCRIPTION
**Pull request by georgibaltiev**
**What this PR does / why we need it**:
This PR is an implementation of Rule 1000 of the Security Hardened Shoot Cluster ruleset for the Garden provider. It evaluates the extensions list in the shoot spec of the cluster and checks for the availability of the extensions provided in the provider configuration.

**Which issue(s) this PR fixes**:
Part of #304 

**Special notes for your reviewer**:

**Release note**:

```feature user
Implementation for rule `1000` from the `security-hardened-shoot-cluster` ruleset for provider `garden`.
```

### CODE REVIEW

File: pkg/provider/garden/ruleset/securityhardenedshoot/rules/1000.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+	if r.Options == nil || len(r.Options.Extensions) == 0 {
+		return rule.Result(r, rule.PassedCheckResult("There are no configured extensions to be evaluated.", rule.NewTarget())), nil
+	}
+
+	if shoot.Spec.Extensions == nil {
+		return rule.Result(r, rule.FailedCheckResult("There are no configured extensions available on the shoot cluster.", rule.NewTarget())), nil
+	}
+
+	var checkResults = []rule.CheckResult{}
```
Comment (AleksandarSavchev) [inline pkg/provider/garden/ruleset/securityhardenedshoot/rules/1000.go:83]:
```suggestion
	var checkResults []rule.CheckResult
```

File: pkg/provider/garden/ruleset/securityhardenedshoot/rules/1000.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+		if extensionTypeIndex < 0 {
+			checkResults = append(checkResults, rule.FailedCheckResult(fmt.Sprintf("Extension %s is not configured for the shoot cluster.", extensionTypeFromOption.Type), rule.NewTarget()))
+		} else {
+			if shoot.Spec.Extensions[extensionTypeIndex].Disabled == nil || !*(shoot.Spec.Extensions[extensionTypeIndex].Disabled) {
+				checkResults = append(checkResults, rule.PassedCheckResult(fmt.Sprintf("Extension %s is enabled for the shoot cluster.", extensionTypeFromOption.Type), rule.NewTarget()))
+			} else {
+				checkResults = append(checkResults, rule.FailedCheckResult(fmt.Sprintf("Extension %s is disabled for the shoot cluster.", extensionTypeFromOption.Type), rule.NewTarget()))
+			}
+		}
```
Comment (AleksandarSavchev) [inline pkg/provider/garden/ruleset/securityhardenedshoot/rules/1000.go:98]:
```suggestion
		if extensionTypeIndex < 0 {
			checkResults = append(checkResults, rule.FailedCheckResult(fmt.Sprintf("Extension type %s is not configured for the shoot cluster.", extensionTypeFromOption.Type), rule.NewTarget()))
		} else {
			if shoot.Spec.Extensions[extensionTypeIndex].Disabled == nil || !*shoot.Spec.Extensions[extensionTypeIndex].Disabled {
				checkResults = append(checkResults, rule.PassedCheckResult(fmt.Sprintf("Extension type %s is enabled for the shoot cluster.", extensionTypeFromOption.Type), rule.NewTarget()))
			} else {
				checkResults = append(checkResults, rule.FailedCheckResult(fmt.Sprintf("Extension type %s is disabled for the shoot cluster.", extensionTypeFromOption.Type), rule.NewTarget()))
			}
		}
```
We can use `switch` for these cases, or as a bare minimum remove nesting.

File: pkg/provider/garden/ruleset/securityhardenedshoot/rules/1000.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+		return rule.Result(r, rule.PassedCheckResult("There are no configured extensions to be evaluated.", rule.NewTarget())), nil
+	}
+
+	if shoot.Spec.Extensions == nil {
+		return rule.Result(r, rule.FailedCheckResult("There are no configured extensions available on the shoot cluster.", rule.NewTarget())), nil
+	}
+
+	var checkResults = []rule.CheckResult{}
+
+	for _, extensionTypeFromOption := range r.Options.Extensions {
```
Comment (AleksandarSavchev) [inline pkg/provider/garden/ruleset/securityhardenedshoot/rules/1000.go:85]:
```suggestion
	for _, extension := range r.Options.Extensions {
```

File: pkg/provider/garden/ruleset/securityhardenedshoot/rules/1000_test.go [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+				{
+					Type: "foo",
+				},
+			},
+			},
+			[]rule.CheckResult{
+				{Status: rule.Failed, Message: "Extension foo is disabled for the shoot cluster.", Target: rule.NewTarget()},
+			},
+		),
+		Entry("should pass when a listed extension has default disabling",
```
Comment (AleksandarSavchev) [inline pkg/provider/garden/ruleset/securityhardenedshoot/rules/1000_test.go:135]:
```suggestion
		Entry("should pass when a listed extension is enabled by default",
```

File: pkg/provider/garden/ruleset/securityhardenedshoot/rules/1000_test.go [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+				{
+					Type: "foo",
+				},
+			},
+			},
+			[]rule.CheckResult{
+				{Status: rule.Passed, Message: "Extension foo is enabled for the shoot cluster.", Target: rule.NewTarget()},
+			},
+		),
+		Entry("should pass when a listed extension has default disabling",
```
Comment (AleksandarSavchev) [inline pkg/provider/garden/ruleset/securityhardenedshoot/rules/1000_test.go:153]:
```suggestion
		Entry("should pass when a listed extension is explicitly enabled",
```

File: pkg/provider/garden/ruleset/securityhardenedshoot/rules/1000.go [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+func (r *Rule1000) Run(ctx context.Context) (rule.RuleResult, error) {
+
+	shoot := &gardencorev1beta1.Shoot{ObjectMeta: v1.ObjectMeta{Name: r.ShootName, Namespace: r.ShootNamespace}}
+	if err := r.Client.Get(ctx, client.ObjectKeyFromObject(shoot), shoot); err != nil {
+		return rule.Result(r, rule.ErroredCheckResult(err.Error(), rule.NewTarget("name", r.ShootName, "namespace", r.ShootNamespace, "kind", "Shoot"))), nil
+	}
+
+	if r.Options == nil || len(r.Options.Extensions) == 0 {
+		return rule.Result(r, rule.PassedCheckResult("There are no configured extensions to be evaluated.", rule.NewTarget())), nil
```
Comment (AleksandarSavchev) [inline pkg/provider/garden/ruleset/securityhardenedshoot/rules/1000.go:76]:
```suggestion
		return rule.Result(r, rule.PassedCheckResult("There are no required extensions", rule.NewTarget())), nil
```

File: pkg/provider/garden/ruleset/securityhardenedshoot/rules/1000.go [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+	}
+
+	if shoot.Spec.Extensions == nil {
+		return rule.Result(r, rule.FailedCheckResult("There are no configured extensions available on the shoot cluster.", rule.NewTarget())), nil
```
Comment (AleksandarSavchev) [inline pkg/provider/garden/ruleset/securityhardenedshoot/rules/1000.go:80]:
```suggestion
		return rule.Result(r, rule.FailedCheckResult("There are no configured extensions on the shoot cluster.", rule.NewTarget())), nil
```

File: pkg/provider/garden/ruleset/securityhardenedshoot/rules/options.go [hunk 8]
```diff
@@ -5,5 +5,5 @@
 package rules
 
 type RuleOption interface {
-	Options2007
+	Options1000 | Options2007
```
Comment (AleksandarSavchev) [inline pkg/provider/garden/ruleset/securityhardenedshoot/rules/options.go:8]:
```suggestion
	Options1000 |
	  Options2007
```

**Review by AleksandarSavchev — CHANGES_REQUESTED** [review]
Thanks!

**Review by AleksandarSavchev — DISMISSED** [review]
/lgtm

File: pkg/provider/garden/ruleset/securityhardenedshoot/rules/1000.go [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+	if err := r.Client.Get(ctx, client.ObjectKeyFromObject(shoot), shoot); err != nil {
+		return rule.Result(r, rule.ErroredCheckResult(err.Error(), rule.NewTarget("name", r.ShootName, "namespace", r.ShootNamespace, "kind", "Shoot"))), nil
+	}
+
+	if r.Options == nil || len(r.Options.Extensions) == 0 {
+		return rule.Result(r, rule.PassedCheckResult("There are no required extensions.", rule.NewTarget())), nil
+	}
+
+	if shoot.Spec.Extensions == nil {
+		return rule.Result(r, rule.FailedCheckResult("There are no configured extensions on the shoot cluster.", rule.NewTarget())), nil
```
Comment (dimityrmirchev) [inline pkg/provider/garden/ruleset/securityhardenedshoot/rules/1000.go:79]:
```suggestion
		return rule.Result(r, rule.FailedCheckResult("There are no configured extensions for the shoot cluster.", rule.NewTarget())), nil
```

File: pkg/provider/garden/ruleset/securityhardenedshoot/rules/1000.go [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+	shoot := &gardencorev1beta1.Shoot{ObjectMeta: v1.ObjectMeta{Name: r.ShootName, Namespace: r.ShootNamespace}}
+	if err := r.Client.Get(ctx, client.ObjectKeyFromObject(shoot), shoot); err != nil {
+		return rule.Result(r, rule.ErroredCheckResult(err.Error(), rule.NewTarget("name", r.ShootName, "namespace", r.ShootNamespace, "kind", "Shoot"))), nil
+	}
+
+	if r.Options == nil || len(r.Options.Extensions) == 0 {
+		return rule.Result(r, rule.PassedCheckResult("There are no required extensions.", rule.NewTarget())), nil
+	}
+
+	if shoot.Spec.Extensions == nil {
```
Comment (dimityrmirchev) [inline pkg/provider/garden/ruleset/securityhardenedshoot/rules/1000.go:78]:
```suggestion
	if len(shoot.Spec.Extensions) == 0 {
```

**Review by dimityrmirchev — CHANGES_REQUESTED** [review]

**Review by dimityrmirchev — APPROVED** [review]
/lgtm

**Pull request merged**